File: packages/async-storage/android/src/main/kotlin/org/asyncstorage/storage/PersistentStorage.kt
package org.asyncstorage.storage

import android.content.Context
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReadableArray
import kotlin.coroutines.CoroutineContext
import kotlin.coroutines.EmptyCoroutineContext
import kotlinx.coroutines.CoroutineName
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.plus
import org.asyncstorage.shared_storage.SharedStorage
import org.asyncstorage.shared_storage.create

private val StorageScope = { name: String -> CoroutineScope(SupervisorJob() + CoroutineName(name)) }

/**
 * todo:
 * - handle exceptions via coroutine exception handler or via try catch
 */
class PersistentStorage(
    ctx: Context,
    dbName: String,
    coroutineContext: CoroutineContext = EmptyCoroutineContext,
) {
    private val scope = StorageScope(dbName) + coroutineContext
    private val db = SharedStorage.create(ctx, dbName)

    fun get(rnKeys: ReadableArray, promise: Promise) {
        scope.launch {
            val keys = rnKeys.toKeyList()
            val result = db.getValues(keys).toRNResults()
            promise.resolve(result)
        }
    }

    fun set(values: ReadableArray, promise: Promise) {
        scope.launch {
            val entries = values.toEntryList()
            val result = db.setValues(entries).toRNResults()
            promise.resolve(result)
        }
    }

    fun remove(keys: ReadableArray, promise: Promise) {
        scope.launch {
            db.removeValues(keys.toKeyList())
            promise.resolve(null)
        }
    }

    fun allKeys(promise: Promise) {
        scope.launch {
            val result = db.getKeys().toRNKeys()
            promise.resolve(result)
        }
    }

    fun clear(promise: Promise) {
        scope.launch {
            db.clear()
            promise.resolve(null)
        }
    }
}

File: packages/async-storage/ios/storage/PersistentStorage.swift
import React
import SharedAsyncStorage

/**
 TODO:
    - handle exceptions
 */
@objc
public class PersistentStorage: NSObject {
    private let db: SharedStorage

    init(databaseName: String) {
        db = SharedStorage.companion.create(context: PlatformContext.Instance(), databaseName: databaseName)
    }

    @objc
    public func get(
        keys: [String],
        resolver: @escaping RCTPromiseResolveBlock,
        rejecter _: @escaping RCTPromiseRejectBlock
    ) {
        Task {
            let result = try await self.db.getValues(keys: keys)
            resolver(result.map { $0.toRNValue() })
        }
    }

    @objc
    public func set(
        values: [[String: String]],
        resolver: @escaping RCTPromiseResolveBlock,
        rejecter _: @escaping RCTPromiseRejectBlock
    ) {
        Task {
            let entries = values.map { entry in Entry.fromRNValue(rnValue: entry) }
            let result = try await self.db.setValues(entries: entries)
            resolver(result.map { entry in entry.toRNValue() })
        }
    }

    @objc
    public func remove(
        keys: [String],
        resolver: @escaping RCTPromiseResolveBlock,
        rejecter _: @escaping RCTPromiseRejectBlock
    ) {
        Task {
            try await self.db.removeValues(keys: keys)
            resolver(nil)
        }
    }

    @objc
    public func allKeys(
        resolver: @escaping RCTPromiseResolveBlock,
        rejecter _: @escaping RCTPromiseRejectBlock
    ) {
        Task {
            let keys = try await self.db.getKeys()
            resolver(keys)
        }
    }

    @objc
    public func clear(
        resolver: @escaping RCTPromiseResolveBlock,
        rejecter _: @escaping RCTPromiseRejectBlock
    ) {
        Task {
            try await self.db.clear()
            resolver(nil)
        }
    }
}

extension Entry {
    // js expects: {key: string, value: string?}
    func toRNValue() -> [String: String?] {
        ["key": key, "value": value]
    }

    static func fromRNValue(rnValue: [String: String]) -> Entry {
        let key: String = rnValue["key"]! // will throw in not there
        let value = rnValue["value"]

        return Entry(key: key, value: value)
    }
}
